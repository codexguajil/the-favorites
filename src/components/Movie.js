import React, { Component } from 'react';

export class Movie extends Component {
  render() {
    const { title, rating, posterImage } = this.props
    return (
      <div>
        <h4>{title}</h4>
        <p>{rating}</p>
        <img src={posterImage} alt='movie poster'/>
      </div>
    )
  }
}

export default Movie
