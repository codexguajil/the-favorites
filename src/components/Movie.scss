@font-face {
  font-family: 'movie-title';
  src: url('./utils/assets/neue-helv/HelveticaNeue-LightCondObl.woff') format('woff2');
}

.movie {
  width: 200px;
  margin: 15px;
  padding-top: 40px;
  position: relative;
  img {
    width: 200px;
  }
  .movie-info {
    background: hsla(0, 0%, 5%, 1);
    color: white;
    visibility: hidden;
    font-family: 'arial';
    height: 200px;
    width: 400px;
    position: fixed;
    // top: 250px;
    // margin-left: 100px;
    // // overflow-y: auto;
    text-overflow: ellipsis;
    z-index: 1;
  }
  &:hover {
    .movie-info {
      visibility: visible;
      // position: fixed;
      // padding: 10px;
      // transform: none;
      // right: auto;
      // bottom: auto;
      // top: auto;
      // left: auto;
      // margin: 0 auto;
      // margin-top: -15px;
      // margin-left: -22px;
    }
  &:after {
    .movie-info {
    }
  }
  }
  // &:after {
  //     border-left: 20px solid transparent;
  //     border-top: 20px solid #efefef;
  //     bottom: -20px;
  //     content: "";
  //     position: absolute;
  //     right: 20px;
  // }
}

.movie-title {
  color: 'black';;
  font-size: 22px;
  padding: 15px;
}

a {
  text-decoration: none;
  color: 'white';
}