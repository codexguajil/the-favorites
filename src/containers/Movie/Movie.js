import React, {Component} from 'react';
import { fetchData } from '../../utils/fetch';
import { addFavorite, deleteFavorite, addMessage } from '../../actions';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import propTypes from 'prop-types';

export class Movie extends Component {
  constructor() {
    super()
    this.state = {
      favorite: false,
      message: ''
    }
  }

  componentDidMount = () => {
    const {favorites, id} = this.props
    favorites.forEach(favorite => {
      if(favorite === id) {
        this.setState({favorite: true})
      }
    })
  }

  validateFavorite = () => {
    const { user_id, id, addMessage } = this.props  
    if(user_id) {
      if(this.props.favorites.includes(id)) {
        this.removeFavorite()
        this.setState({favorite: false})
      } else {
        this.addFavorite()
        this.setState({favorite: true})
      }
    } else {
      addMessage('Please log in to add to favorites.')
      setTimeout(() => {
        addMessage('')
      }, 3000)
    }
  }

  removeFavorite = async () => {
    const { id, user_id } = this.props
    const url = `http://localhost:3000/api/users/${user_id}/favorites/${id}`;
    const userOptionObject = {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json"
      }
    }
    await fetchData(url, userOptionObject);
    this.props.deleteFavorite(id)
  }

  addFavorite = async (e) => {
    const { id, title, rating, user_id, posterImage, synopsis, releaseDate } = this.props
    const url = "http://localhost:3000/api/users/favorites/new"
    const movie = {
      movie_id: id, 
      user_id: user_id, 
      title: title, 
      poster_path: posterImage, 
      release_date: releaseDate, 
      vote_average: rating, 
      overview: synopsis
    }
    try {
      const addedFavorite = await fetchData(url, {
        method: "POST", 
        body: JSON.stringify(movie),
        headers: {
          "Content-Type": "application/json"
        }
      })
      // console.log(addedFavorite)
      this.props.addFavoriteToState(id)
    } catch(error) {
      console.log(error.message)
    }
  }

  render() {
    const {favorite} = this.state
    const { id, title, rating, posterImage, synopsis, type } = this.props
    const image = 'https://image.tmdb.org/t/p/w500'+ posterImage
    return (
      <div className="movie">
        <div className="movie-info">
        {title}
        {rating}
        {synopsis}
        </div>
        {/* <h4 className="movie-title">{title}</h4> */}
        {/* <p classsName="rating">{rating}</p> */}
        <Link to={`/${type}/${id}`}>
          <img src={image} alt='movie poster'/> 
        </Link>
        <div onClick={this.validateFavorite} className={favorite ? 'fav-true': 'fav-false'}></div>
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  user_id: state.user.id,
  favorites: state.favorites
})

export const mapDispatchToProps = (dispatch) => ({
  addFavoriteToState: (favorite) => dispatch(addFavorite(favorite)),
  deleteFavorite: (favorite) => dispatch(deleteFavorite(favorite)),
  addMessage: (message) => dispatch(addMessage(message))
})

export default connect(mapStateToProps, mapDispatchToProps)(Movie)

Movie.propTypes = {
  user_id: propTypes.number,
  favorites: propTypes.array.isRequired,
  addFavoriteToState: propTypes.func.isRequired,
  deleteFavorite: propTypes.func.isRequired,
  addMessage: propTypes.func.isRequired
}
