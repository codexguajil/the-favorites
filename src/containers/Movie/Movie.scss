@font-face {
  font-family: 'movie-title';
  src: url('./utils/assets/neue-helv/HelveticaNeue-LightCondObl.woff') format('woff2');
}

.movie {
  width: 200px;
  margin: 15px;
  padding-top: 40px;
  position: relative;
  img {
    opacity: .75;
    width: 200px;
  }
  .movie-info {
    background: hsla(0, 0%, 5%, 1);
    color: white;
    visibility: hidden;
    font-family: 'arial';
    height: 200px;
    width: 400px;
    position: fixed;
    text-overflow: ellipsis;
    z-index: 1;
  }
  &:hover {
    img {
      opacity: 1;
      box-shadow: 1px 1px 12px 3px hsla(0, 0%, 5%, 1);
    }
  }
}

.fav-false {
  height: 65px;
  width: 65px;
  position: absolute;
  top: 24px;
  left: -19px;
  background: url('./utils/assets/false.png');
  background-repeat: no-repeat;
}

.fav-true {
  height: 65px;
  width: 65px;
  position: absolute;
  top: 24px;
  left: -19px;
  background: url('./utils/assets/true.png')
}

.movie-title {
  color: 'black';;
  font-size: 22px;
  padding: 15px;
}

a {
  text-decoration: none;
  color: 'white';
}