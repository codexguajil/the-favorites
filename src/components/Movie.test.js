import React from 'react'
import Movie from './Movie'
import { shallow } from 'enzyme'

describe('Movie', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(
      <Movie />
    )

    expect(wrapper).toMatchSnapshot()
  })
})

