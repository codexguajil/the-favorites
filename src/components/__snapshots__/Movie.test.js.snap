// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie should match snapshot 1`] = `
<div>
  <h4 />
  <p />
  <img
    alt="movie poster"
  />
</div>
`;
